src/lib.rs

```diff
@@ -221,84 +203,103 @@ impl Renderer {
+        device.SetRenderState(D3DRS_ALPHATESTENABLE, FALSE).unwrap();
+        device.SetRenderState(D3DRS_BLENDOP, D3DBLENDOP_ADD.0 as u32).unwrap();
+        device.SetRenderState(D3DRS_SRCBLEND, D3DBLEND_SRCALPHA.0 as u32).unwrap();
+        device.SetRenderState(D3DRS_DESTBLEND, D3DBLEND_INVSRCALPHA.0 as u32).unwrap();
+        device.SetRenderState(D3DRS_SCISSORTESTENABLE, TRUE).unwrap();
+        device.SetRenderState(D3DRS_SHADEMODE, D3DSHADE_GOURAUD.0 as u32).unwrap();
+        device.SetRenderState(D3DRS_FOGENABLE, FALSE).unwrap();
+        device.SetTextureStageState(0, D3DTSS_COLOROP, D3DTOP_MODULATE.0 as u32).unwrap();
+        device.SetTextureStageState(0, D3DTSS_COLORARG1, D3DTA_TEXTURE).unwrap();
+        device.SetTextureStageState(0, D3DTSS_COLORARG2, D3DTA_DIFFUSE).unwrap();
+        device.SetTextureStageState(0, D3DTSS_ALPHAOP, D3DTOP_MODULATE.0 as u32).unwrap();
+        device.SetTextureStageState(0, D3DTSS_ALPHAARG1, D3DTA_TEXTURE).unwrap();
+        device.SetTextureStageState(0, D3DTSS_ALPHAARG2, D3DTA_DIFFUSE).unwrap();
+        device.SetSamplerState(0, D3DSAMP_MINFILTER, D3DTEXF_LINEAR.0 as u32).unwrap();
+        device.SetSamplerState(0, D3DSAMP_MAGFILTER, D3DTEXF_LINEAR.0 as u32).unwrap();
 
         let l = draw_data.display_pos[0] + 0.5;
         let r = draw_data.display_pos[0] + draw_data.display_size[0] + 0.5;
         let t = draw_data.display_pos[1] + 0.5;
         let b = draw_data.display_pos[1] + draw_data.display_size[1] + 0.5;
         let mat_projection = D3DMATRIX {
-            m: [
-                [2.0 / (r - l), 0.0, 0.0, 0.0],
-                [0.0, 2.0 / (t - b), 0.0, 0.0],
-                [0.0, 0.0, 0.5, 0.0],
-                [(l + r) / (l - r), (t + b) / (b - t), 0.5, 1.0],
-            ],
+            Anonymous: D3DMATRIX_0 {
+                m: [
+                    2.0 / (r - l),
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    2.0 / (t - b),
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.5,
+                    0.0,
+                    (l + r) / (l - r),
+                    (t + b) / (b - t),
+                    0.5,
+                    1.0,
+                ],
+            },
         };
-        device.SetTransform(D3DTS_WORLD, &MAT_IDENTITY);
-        device.SetTransform(D3DTS_VIEW, &MAT_IDENTITY);
-        device.SetTransform(D3DTS_PROJECTION, &mat_projection);
+
+        device.SetTransform(D3DTRANSFORMSTATETYPE(0), &MAT_IDENTITY).unwrap();
+        device.SetTransform(D3DTS_VIEW, &MAT_IDENTITY).unwrap();
+        device.SetTransform(D3DTS_PROJECTION, &mat_projection).unwrap();
     }
 
     unsafe fn lock_buffers<'v, 'i>(
-        vb: &'v mut IDirect3DVertexBuffer9,
-        ib: &'i mut IDirect3DIndexBuffer9,
+        vb: &IDirect3DVertexBuffer9,
+        ib: &IDirect3DIndexBuffer9,
```

Comment:
```suggestion
        vb: &'v mut IDirect3DVertexBuffer9,
        ib: &'i mut IDirect3DIndexBuffer9,
```

Reply:
I'm not sure about this change, the compiler does not complain about the mutability or the lifecycle, so it should be fine as is?

Reply:
It doesn't but only because it can't reason about the content here. Regarding the lifetimes, if we don't specify the inputs one could call `lock_buffers` in a way that would yield `'static` lifetimes which is very much wrong, that is if we don't link the lifetimes of the output slices with any input the lifetimes are arbitrary, but they are very much only alive as long as the buffers are. Regarding the mutability, while the Lock/Unlock functions only need immtuable access, the returned slices are mutable slices, so if we leave those input parameters as immutable references one could call this function multiple times and get more than one mutable reference to the same data which is undefined behaviour in rust.

This is why unsafe code can be tricky when dealing with references and lifetimes, as the compiler can't reason about it

Reply:
Alright makes sense, thanks for the explanation. Fixed.